File: src/gestorAplicacion/usuario/Meta.java
package gestorAplicacion.usuario;

import gestorAplicacion.administrador.Utils;
import gestorAplicacion.administrador.Validador;
import gestorAplicacion.economia.*;

import java.io.Serializable;
import java.time.LocalDate;

public class Meta implements Serializable, Abonable, Contable {

	private static final long serialVersionUID = 659116063038663746L;

	private Usuario usuario;

	private String nombre;
	private boolean cumplida;
	private LocalDate fechaCumplimiento;
	private LocalDate fechaInicio;

	private double objetivo;

	private Divisa divisa;

	private double saldo = 0;

	public Meta(Usuario usuario, String nombre, LocalDate fechaInicio, double objetivo, Divisa divisa) {
		this.nombre = nombre;
		this.cumplida = false;
		this.fechaCumplimiento = null;
		this.fechaInicio = fechaInicio;
		this.objetivo = objetivo;
		this.divisa = divisa;
		this.usuario = usuario;
		this.saldo = 0;
	}

	public String getNombre() {
		return nombre;
	}
	
	public void setNombre(String nombre) {
		this.nombre = nombre;
	}
	
	public boolean isCumplida() {
		return cumplida;
	}
	
	public void setCumplida(boolean cumplida) {
		this.cumplida = cumplida;
	}


	public double getObjetivo() {
		return objetivo;
	}

	public void setObjetivo(double objetivo) {
		this.objetivo = objetivo;
		this.metaCumplida();
	}
	private void metaCumplida(){
		if(this.saldo>= this.objetivo ){
			System.err.println("FELICIDADES HAS CUMPLIDO TU META "+ this.nombre.toUpperCase());
			System.out.println("Escoge un Bolsillo al cual enviar el dinero para que lo puedas usar ("+this.saldo+" "+this.divisa+"): ");
			this.usuario.listarBolsillos();
			int option = Validador.validarEntradaInt(this.usuario.getBolsillos().size(), true, 1, true) -1;
			Bolsillo bolsillo = this.usuario.getBolsillos().get(option);
			double[] nuevoSaldo = this.divisa.ConvertToDivisa(this.saldo,bolsillo.getDivisa());
			this.saldo = 0;
			this.cumplida = true;
			this.fechaCumplimiento = LocalDate.now();
			Ingreso ingreso = new Ingreso(nuevoSaldo[0], LocalDate.now(), true,null, null,bolsillo,this.divisa,bolsillo.getDivisa());
			usuario.nuevoIngreso(ingreso);
			System.out.println("Nuevo saldo en el bolsillo de: "+ nuevoSaldo[0]+ " "+ bolsillo.getDivisa());
			System.out.println("TRM usada de: "+ nuevoSaldo[1]);
		}
	}
	public Divisa getDivisa() {
		return divisa;
	}

	public void setDivisa(Divisa divisa) {
		this.divisa = divisa;
	}

	public double getSaldo() {
		return saldo;
	}

	public void setSaldo(double saldo) {
		this.saldo = saldo;
	}

	public Usuario getUsuario() {
		return usuario;
	}

	public void setUsuario(Usuario usuario) {
		this.usuario = usuario;
	}

	public LocalDate getFechaCumplimiento() {
		return fechaCumplimiento;
	}

	public void setFechaCumplimiento(LocalDate fechaCumplimiento) {
		this.fechaCumplimiento = fechaCumplimiento;
	}

	public LocalDate getFechaInicio() {
		return fechaInicio;
	}

	public void setFechaInicio(LocalDate fechaInicio) {
		this.fechaInicio = fechaInicio;
	}

	@Override
	public Salida abonar(double monto, Cuenta origen) {
		this.saldo += monto;
		metaCumplida();
		return null;
	}
}

File: Python/gestorAplicacion/usuario/Meta.py
from gestorAplicacion.economia import Abonable, Contable, Salida, Ingreso
from datetime import date, time, datetime

class Meta(Abonable, Contable):

    def __init__(self, usuario, nombre, fechaInicio, objetivo, divisa):
        
        self._nombre = nombre
        self._cumplida = False
        self._fechaCumplimiento = None
        self._fechaInicio = fechaInicio
        self._objetivo = objetivo
        self._divisa = divisa
        self._usuario = usuario
        self._saldo = 0

    def getNombre(self):
        return self._nombre

    def setNombre(self, nombre):
        self._nombre = nombre

    def isCumplida(self):
        return self._cumplida

    def setCumplida(self, cumplida):
        self._cumplida = cumplida


    def getObjetivo(self):
        return self._objetivo

    def setObjetivo(self, objetivo):
        bol = [False for _ in range(2)]
        if not self._cumplida:
            self._objetivo = objetivo
            bol[1] = self.metaCumplida()
            bol[0] = True
        else:
            bol[0] = False
            bol[1] = False
        return bol

    def metaCumplida(self):
        return self._saldo>= self._objetivo


    def getDivisa(self):
        return self._divisa

    def setDivisa(self, divisa):
        self._divisa = divisa

    def getSaldo(self):
        return self._saldo

    def setSaldo(self, saldo):
        self._saldo = saldo

    def getUsuario(self):
        return self._usuario

    def setUsuario(self, usuario):
        self._usuario = usuario

    def getFechaCumplimiento(self):
        return self._fechaCumplimiento

    def setFechaCumplimiento(self, fechaCumplimiento):
        self._fechaCumplimiento = fechaCumplimiento

    def getFechaInicio(self):
        return self._fechaInicio

    def setFechaInicio(self, fechaInicio):
        self._fechaInicio = fechaInicio

    def abonar(self, monto, origen):
        if not self._cumplida:
            monto2 = origen.getDivisa().ConvertToDivisa(monto,self._divisa)
            salida = Salida(monto2[0],monto, date.today(), origen, None, origen.getDivisa(), self._divisa)
            retirado = self._usuario.nuevaSalida(salida)
            if not retirado:
                return None
            self._saldo += monto2[0]
            return salida
        return None

    def terminar(self, cuenta):
        nuevoSaldo = self.getDivisa().ConvertToDivisa(self._saldo, cuenta.getDivisa())
        self._saldo = 0
        self._cumplida = True
        self._fechaCumplimiento = date.today()
        ingreso = Ingreso(nuevoSaldo[0], self._saldo, date.today(), True, None, None, cuenta, self._divisa, cuenta.getDivisa())
        self._usuario.nuevoIngreso(ingreso)
        return ingreso
